src/dbnode/persist/fs/msgpack/stream_with_digest.go

```diff
@@ -0,0 +1,114 @@
+package msgpack
+
+import (
+	"errors"
+	"hash"
+	"hash/adler32"
+)
+
+var (
+	// errChecksumMismatch returned when the calculated checksum doesn't match the stored checksum
+	errChecksumMismatch = errors.New("calculated checksum doesn't match stored checksum")
+)
+
+// DecoderStreamWithDigest calculates the digest as it processes a decoder stream.
+type DecoderStreamWithDigest interface {
```

Comment:
I would actually just move this to the "decoder" and always wrap the incoming stream.

That way we don't need to conditionally check whether the stream is this type or not from within the decoder itself, we can just blindly wrap the stream that was passed into "decoder.Reset(...)".

Reply:
> I would actually just move this to the "decoder" 

You mean move this interface and impl to the decoder.go file?

> and always wrap the incoming stream.

This makes sense to me. My concern here was that we'd then always be calculating the digest, but that's probably not that big of a deal and a bit of a premature optimization.